sdk/test/opentelemetry/sdk/trace/span_test.rb

```diff
@@ -7,4 +7,220 @@
+    end
+
+    it 'does not set an attribute if span is ended' do
+      span.finish
+      span.set_attribute('no', 'set')
+      span.to_proto[:attributes].must_be_nil
+    end
+
+    it 'counts attributes' do
+      span.set_attribute('old', 'oldbar')
+      span.set_attribute('foo', 'bar')
+      span.to_proto[:total_recorded_attributes].must_equal(2)
+    end
+  end
+
+  describe '#add_event' do
+    it 'add a named event' do
+      span.add_event(name: 'added')
+      events = span.to_proto[:events]
+      events.size.must_equal(1)
+      events.first.name.must_equal('added')
+    end
+
+    it 'add event with attributes' do
+      attrs = { 'foo' => 'bar' }
+      span.add_event(name: 'added', attributes: attrs)
+      events = span.to_proto[:events]
+      events.size.must_equal(1)
+      events.first.attributes.must_equal(attrs)
+    end
+
+    it 'add event with timestamp' do
+      ts = Time.now
+      span.add_event(name: 'added', timestamp: ts)
+      events = span.to_proto[:events]
+      events.size.must_equal(1)
+      events.first.timestamp.must_equal(ts)
+    end
+
+    it 'add an event as event formatter' do
+      span.add_event { Event.new(name: 'c', attributes: nil, timestamp: Time.now) }
+      events = span.to_proto[:events]
+      events.size.must_equal(1)
+      events.first.name.must_equal('c')
+    end
+
+    it 'does not add an event if span is ended' do
+      span.finish
+      span.add_event(name: 'will_not_be_added')
+      span.to_proto[:events].must_be_nil
+    end
+  end
+
+  describe '#status=' do
+    it 'sets the status' do
+      span.status = Status.new(1, description: 'cancelled')
+      span.to_proto[:status].description.must_equal('cancelled')
+    end
+
+    it 'does not set the status if span is ended' do
+      span.finish
+      span.status = Status.new(1, description: 'cancelled')
+      span.to_proto[:status].must_be_nil
+    end
+  end
+
+  describe '#name=' do
+    it 'sets the name' do
+      span.name = 'new_name'
+      span.to_proto[:name].must_equal('new_name')
+    end
+
+    it 'does not set the name if span is ended' do
+      span.finish
+      span.name = 'new_name'
+      span.to_proto[:name].must_equal('name')
+    end
+  end
+
+  describe '#finish' do
+    it 'returns itself' do
+      span.finish.must_equal(span)
+    end
+
+    it 'sets the end timestamp' do
+      span.finish
+      span.to_proto[:end_timestamp].wont_be_nil
+    end
+
+    it 'calls the span processor #on_end callback' do
+      mock_span_processor.expect(:on_start, nil) do |_|
+        true
+      end
+      span = Span.new(context, 'name', SpanKind::INTERNAL, nil, trace_config,
+                      mock_span_processor, nil, nil, nil, Time.now)
+      mock_span_processor.expect :on_end, nil, [span]
+      span.finish
+      mock_span_processor.verify
+    end
+
+    it 'marks the span as ended' do
+      span.finish
+      span.instance_variable_get(:@ended).must_equal(true)
```

Comment:
Rather than add `ended` to the `to_proto` export this is left as `instance_variable_get`